File: system-design/oop-patterns/02-observer-pattern/full/ts/weather-station/WeatherStation.ts
import type { IObservable } from './IObservable';
import type { IObserver } from './IObserver';

export class WeatherStation implements IObservable {
  private observers: IObserver[] = [];
  private temperature = 0;

  add(observer: IObserver): void {
    this.observers.push(observer);
  }

  remove(observer: IObserver): void {
    this.observers = this.observers.filter((o) => o !== observer);
  }

  notify(): void {
    for (const observer of this.observers) {
      observer.update();
    }
  }

  getTemperature(): number {
    return this.temperature;
  }

  setTemperature(temperature: number): void {
    this.temperature = temperature;
    this.notify();
  }
}

File: system-design/oop-patterns/02-observer-pattern/full/java/weatherstation/WeatherStation.java
package weatherstation;

import java.util.ArrayList;
import java.util.List;

public class WeatherStation implements IObservable {
    private final List<IObserver> observers = new ArrayList<>();
    private int temperature = 0;

    public void add(IObserver observer) {
        observers.add(observer);
    }

    public void remove(IObserver observer) {
        observers.remove(observer);
    }

    public void notifyObservers() {
        for (IObserver observer : observers) {
            observer.update();
        }
    }

    public int getTemperature() {
        return temperature;
    }

    public void setTemperature(int temperature) {
        this.temperature = temperature;
        notifyObservers();
    }
}
